Build error in cpu_blas.cpp

I am trying to build the v3.6 branch on Debian Buster. However, I get the error message:

```
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp: In function ‘void (* opencl::cpu::gemm_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, opencl::cpu::scale_type<T>, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::scale_type<T>, opencl::cpu::ptr_type<T>, blasint) [with T = cl_float2; opencl::cpu::gemm_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const float*, const float*, int, const float*, int, const float*, float*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; opencl::cpu::scale_type<T> = const float*; opencl::cpu::cptr_type<T> = const float*; opencl::cpu::ptr_type<T> = float*]’:
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:116:16: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint)’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)’} to ‘opencl::cpu::gemm_func_def<cl_float2>’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const float*, const float*, int, const float*, int, const float*, float*, int)’} [-fpermissive]
         return &cblas_##PREFIX##FUNC;           \
                ^~~~~~~
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:122:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemm, cfloat, c)
 ^~~~~~~~~
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp: In function ‘void (* opencl::cpu::gemm_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, opencl::cpu::scale_type<T>, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::scale_type<T>, opencl::cpu::ptr_type<T>, blasint) [with T = cl_double2; opencl::cpu::gemm_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const double*, const double*, int, const double*, int, const double*, double*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; opencl::cpu::scale_type<T> = const double*; opencl::cpu::cptr_type<T> = const double*; opencl::cpu::ptr_type<T> = double*]’:
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:116:16: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, blasint, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint)’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)’} to ‘opencl::cpu::gemm_func_def<cl_double2>’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, CBLAS_TRANSPOSE, int, int, int, const double*, const double*, int, const double*, int, const double*, double*, int)’} [-fpermissive]
         return &cblas_##PREFIX##FUNC;           \
                ^~~~~~~
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:123:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemm, cdouble, z)
 ^~~~~~~~~
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp: In function ‘void (* opencl::cpu::gemv_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, opencl::cpu::scale_type<T>, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::scale_type<T>, opencl::cpu::ptr_type<T>, blasint) [with T = cl_float2; opencl::cpu::gemv_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const float*, const float*, int, const float*, int, const float*, float*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; opencl::cpu::scale_type<T> = const float*; opencl::cpu::cptr_type<T> = const float*; opencl::cpu::ptr_type<T> = float*]’:
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:116:16: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint)’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)’} to ‘opencl::cpu::gemv_func_def<cl_float2>’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const float*, const float*, int, const float*, int, const float*, float*, int)’} [-fpermissive]
         return &cblas_##PREFIX##FUNC;           \
                ^~~~~~~
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:128:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemv, cfloat, c)
 ^~~~~~~~~
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp: In function ‘void (* opencl::cpu::gemv_func())(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, opencl::cpu::scale_type<T>, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::cptr_type<T>, blasint, opencl::cpu::scale_type<T>, opencl::cpu::ptr_type<T>, blasint) [with T = cl_double2; opencl::cpu::gemv_func_def<T> = void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const double*, const double*, int, const double*, int, const double*, double*, int); CBLAS_ORDER = CBLAS_ORDER; CBLAS_TRANSPOSE = CBLAS_TRANSPOSE; blasint = int; opencl::cpu::scale_type<T> = const double*; opencl::cpu::cptr_type<T> = const double*; opencl::cpu::ptr_type<T> = double*]’:
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:116:16: error: invalid conversion from ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, blasint, blasint, const void*, const void*, blasint, const void*, blasint, const void*, void*, blasint)’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const void*, const void*, int, const void*, int, const void*, void*, int)’} to ‘opencl::cpu::gemv_func_def<cl_double2>’ {aka ‘void (*)(CBLAS_ORDER, CBLAS_TRANSPOSE, int, int, const double*, const double*, int, const double*, int, const double*, double*, int)’} [-fpermissive]
         return &cblas_##PREFIX##FUNC;           \
                ^~~~~~~
/home/ralf/devel/arrayfire/src/backend/opencl/cpu/cpu_blas.cpp:129:1: note: in expansion of macro ‘BLAS_FUNC’
 BLAS_FUNC(gemv, cdouble, z)
 ^~~~~~~~~
make[2]: *** [src/backend/opencl/CMakeFiles/afopencl.dir/build.make:2183: src/backend/opencl/CMakeFiles/afopencl.dir/cpu/cpu_blas.cpp.o] Error 1
```

This looks very similar to #2321, but my set-up is different. All packages are installed via `apt` from the Debian Buster repositories, giving me gcc 8.3.0 and OpenBLAS 0.3.5. The solution suggested in #2321 included replacing `/usr/include/x86_64-linux-gnu/cblas.h` with `/usr/include/x86_64-linux-gnu/cblas-netlib.h`. However, I do not think this is correct, since the former file points to `/usr/include/x86_64-linux-gnu/cblas-openblas.h` via `/etc/alternatives/cblas.h-x86_64-linux-gnu`, which appears maore sensible to me when using OpenBLAS instead of Netlib.

Any advice?

--- a/src/backend/opencl/cpu/cpu_blas.cpp
+++ b/src/backend/opencl/cpu/cpu_blas.cpp
@@ -113,7 +113,7 @@ using gemv_func_def = void (*)(const CBLAS_ORDER, const CBLAS_TRANSPOSE,
 #define BLAS_FUNC(FUNC, TYPE, PREFIX)           \
     template<>                                  \
     FUNC##_func_def<TYPE> FUNC##_func<TYPE>() { \
-        return &cblas_##PREFIX##FUNC;           \
+        return (FUNC##_func_def<TYPE>)&cblas_##PREFIX##FUNC; \
     }
 
 BLAS_FUNC_DEF(gemm)

